examples/tutorials/api/irfs.py

```diff
@@ -0,0 +1,289 @@
+######################################################################
+# While Gammapy does not have inbuilt classes for supporting asymmetric
+# IRFs (except for `Background3D`), custom classes can be created. For
+# this to work correctly with the `MapDatasetMaker`, only variations
+# with `fov_lon` and `fov_lat` can be allowed.
+#
+# Analytic models for the PSF, or anisotropic PSFs (ie, asymmetry around
+# the source position) is not correctly supported during the data
+# reduction
+#
+
+import numpy as np
+import scipy.special
+import astropy.units as u
+from astropy.coordinates import SkyCoord
+from gammapy.irf import *
+from gammapy.irf.io import COMMON_IRF_HEADERS, IRF_DL3_HDU_SPECIFICATION
+from gammapy.makers.utils import *
+from gammapy.maps import MapAxes, MapAxis, WcsGeom
+
+######################################################################
+# Effective Area
+# --------------
+#
+
+
+class EffectiveArea3D(IRF):
+    tag = "aeff_3d"
+    required_axes = ["energy_true", "fov_lon", "fov_lat"]
+    default_unit = u.m**2
+
+
+energy_axis = MapAxis.from_energy_edges(
+    [0.1, 0.3, 1.0, 3.0, 10.0] * u.TeV, name="energy_true"
+)
+
+nbin = 7
+fov_lon_axis = MapAxis.from_edges([-1.5, -0.5, 0.5, 1.5] * u.deg, name="fov_lon")
+fov_lat_axis = MapAxis.from_edges([-1.5, -0.5, 0.5, 1.5] * u.deg, name="fov_lat")
+
+data = np.ones((4, 3, 3))
+for i in range(1, 4):
+    data[i] = data[i - 1] * 1.5
+
+aeff_3d = EffectiveArea3D(
+    [energy_axis, fov_lon_axis, fov_lat_axis], data=data, unit=u.m**2
+)
+print(aeff_3d)
+
+res = aeff_3d.evaluate(
+    fov_lon=[-0.5, 0.8] * u.deg,
+    fov_lat=[-0.5, 1.0] * u.deg,
+    energy_true=[0.2, 8.0] * u.TeV,
+)
+print(res)
+
+
+######################################################################
+# Serialisation
+# ~~~~~~~~~~~~~
+#
+
+IRF_DL3_HDU_SPECIFICATION["aeff_3d"] = {
+    "extname": "EFFECTIVE AREA",
+    "column_name": "EFFAREA",
+    "mandatory_keywords": {
+        **COMMON_IRF_HEADERS,
+        "HDUCLAS2": "EFF_AREA",
+        "HDUCLAS3": "FULL-ENCLOSURE",  # added here to have HDUCLASN in order
+        "HDUCLAS4": "AEFF_3D",
+    },
+}
+
+aeff_3d.write("test_aeff3d.fits", overwrite=True)
+
+aeff_new = EffectiveArea3D.read("test_aeff3d.fits")
+print(aeff_new)
+
+
+######################################################################
+# Create exposure map
+# ~~~~~~~~~~~~~~~~~~~
+#
+
+axis = MapAxis.from_energy_bounds(0.1 * u.TeV, 10 * u.TeV, 6, name="energy_true")
+pointing = SkyCoord(2, 1, unit="deg")
+geom = WcsGeom.create(npix=(4, 3), binsz=2, axes=[axis], skydir=pointing)
+
+print(geom)
+
+exposure_map = make_map_exposure_true_energy(
+    pointing=pointing, livetime="42 h", aeff=aeff_3d, geom=geom
+)
+
+
+######################################################################
+# Energy Dispersion
+# -----------------
+#
+
+
+class EnergyDispersion3D(IRF):
+    tag = "edisp_3d"
+    required_axes = ["energy_true", "migra", "fov_lon", "fov_lat"]
+    default_unit = u.one
+
+    @classmethod
```

Comment:
you might not expose this in a tutorial as this is quite complex. Showing a simple diagonal one would be enough.

Reply:
Indeed! This is just to show that most of the functions can be easily generalised to the 3D case.